From: Miklos Szeredi <mszeredi@suse.cz>
Date: Tue, 21 Feb 2012 16:13:12 +0800
Subject: [PATCH 5/11] vfs: introduce clone_private_mount()
Patch-mainline: in house patch

Overlayfs needs a private clone of the mount, so create a function for
this and export to modules.

Signed-off-by: Miklos Szeredi <mszeredi@suse.cz>
Signed-off-by: Robin Dong <sanbai@taobao.com>
Signed-off-by: Coly Li <bosong.ly@taobao.com>
---
 fs/namespace.c        |   18 ++++++++++++++++++
 include/linux/mount.h |    3 +++
 2 files changed, 21 insertions(+), 0 deletions(-)

Index: linux-2.6.32-279.2.1.el5/fs/namespace.c
===================================================================
--- linux-2.6.32-279.2.1.el5.orig/fs/namespace.c
+++ linux-2.6.32-279.2.1.el5/fs/namespace.c
@@ -1274,6 +1274,24 @@ void drop_collected_mounts(struct vfsmou
 	release_mounts(&umount_list);
 }
 
+struct vfsmount *clone_private_mount(struct path *path)
+{
+	struct vfsmount *old_mnt = path->mnt;
+	struct vfsmount *new_mnt;
+
+	if (IS_MNT_UNBINDABLE(old_mnt))
+		return ERR_PTR(-EINVAL);
+
+	down_read(&namespace_sem);
+	new_mnt = clone_mnt(old_mnt, path->dentry, CL_PRIVATE);
+	up_read(&namespace_sem);
+	if (!new_mnt)
+		return ERR_PTR(-ENOMEM);
+
+	return new_mnt;
+}
+EXPORT_SYMBOL_GPL(clone_private_mount);
+
 static void cleanup_group_ids(struct vfsmount *mnt, struct vfsmount *end)
 {
 	struct vfsmount *p;
Index: linux-2.6.32-279.2.1.el5/include/linux/mount.h
===================================================================
--- linux-2.6.32-279.2.1.el5.orig/include/linux/mount.h
+++ linux-2.6.32-279.2.1.el5/include/linux/mount.h
@@ -111,6 +111,9 @@ static inline void mntput(struct vfsmoun
 extern struct vfsmount *do_kern_mount(const char *fstype, int flags,
 				      const char *name, void *data);
 
+struct path;
+extern struct vfsmount *clone_private_mount(struct path *path);
+
 struct file_system_type;
 extern struct vfsmount *vfs_kern_mount(struct file_system_type *type,
 				      int flags, const char *name,
